### PR TITLE
Refactor optional imports

### LINKED ISSUE #215: Avoid importing optional dependencies at module level
## 🐛 Bug

LitData has optional dependencies such as torchvision. If user does not use any torchvision related features in LitData, LitData will still attempt to import it. This has two major drawbacks:

1. Affects import time
2. If the dependency has an error in its import path, or has incompatibility with other packages, the user gets an import error even though they don't use the package at all.

### To Reproduce

1. Install an incompatible combination of torch and torchvision.
2. `import litdata`

You'll get an error like this (not exactly the same but it's the same root issue):
```
Traceback (most recent call last):
  File "/teamspace/studios/this_studio/main.py", line 5, in <module>
    from litdata import map
  File "/home/zeus/miniconda3/envs/cloudspace/lib/python3.10/site-packages/litdata/__init__.py", line 16, in <module>
    from litdata.processing.functions import map, optimize, walk
  File "/home/zeus/miniconda3/envs/cloudspace/lib/python3.10/site-packages/litdata/processing/functions.py", line 27, in <module>
    from litdata.processing.data_processor import DataChunkRecipe, DataProcessor, DataTransformRecipe
  File "/home/zeus/miniconda3/envs/cloudspace/lib/python3.10/site-packages/litdata/processing/data_processor.py", line 44, in <module>
    from litdata.processing.readers import BaseReader, StreamingDataLoaderReader
  File "/home/zeus/miniconda3/envs/cloudspace/lib/python3.10/site-packages/litdata/processing/readers.py", line 20, in <module>
    from litdata.streaming.dataloader import StreamingDataLoader
  File "/home/zeus/miniconda3/envs/cloudspace/lib/python3.10/site-packages/litdata/streaming/__init__.py", line 14, in <module>
    from litdata.streaming.cache import Cache
  File "/home/zeus/miniconda3/envs/cloudspace/lib/python3.10/site-packages/litdata/streaming/cache.py", line 21, in <module>
    from litdata.streaming.item_loader import BaseItemLoader, Interval
  File "/home/zeus/miniconda3/envs/cloudspace/lib/python3.10/site-packages/litdata/streaming/item_loader.py", line 28, in <module>
    from litdata.streaming.serializers import Serializer
  File "/home/zeus/miniconda3/envs/cloudspace/lib/python3.10/site-packages/litdata/streaming/serializers.py", line 54, in <module>
    if _TORCH_VISION_AVAILABLE:
  File "/home/zeus/miniconda3/envs/cloudspace/lib/python3.10/site-packages/litdata/imports.py", line 111, in __bool__
    self._check_requirement()
  File "/home/zeus/miniconda3/envs/cloudspace/lib/python3.10/site-packages/litdata/imports.py", line 105, in _check_requirement
    self.available = module_available(module)
  File "/home/zeus/miniconda3/envs/cloudspace/lib/python3.10/site-packages/litdata/imports.py", line 58, in module_available
    importlib.import_module(module_path)
  File "/home/zeus/miniconda3/envs/cloudspace/lib/python3.10/importlib/__init__.py", line 126, in import_module
    return _bootstrap._gcd_import(name[level:], package, level)
  File "/home/zeus/miniconda3/envs/cloudspace/lib/python3.10/site-packages/torchvision/__init__.py", line 6, in <module>
    from torchvision import _meta_registrations, datasets, io, models, ops, transforms, utils
  File "/home/zeus/miniconda3/envs/cloudspace/lib/python3.10/site-packages/torchvision/_meta_registrations.py", line 26, in <module>
    def meta_roi_align(input, rois, spatial_scale, pooled_height, pooled_width, sampling_ratio, aligned):
  File "/home/zeus/miniconda3/envs/cloudspace/lib/python3.10/site-packages/torchvision/_meta_registrations.py", line 18, in wrapper
    if torchvision.extension._has_ops():
AttributeError: partially initialized module 'torchvision' has no attribute 'extension' (most likely due to a circular import)
```

### Expected behavior

Optional dependencies don't get imported when not used. Move import checks like 

```
if _XYZ_AVAILABLE:
    import xyz
```

from the module-level to the specific code section where it is used.

### PR DESCRIPTION
**Pull request by awaelchli**
Fixes #215 

TLDR: Moves optional imports and checks to local imports. This avoids import issues for unrelated features.
See example in #215 why importing optional dependencies at module-level can be problematic.

### CODE REVIEW

File: src/litdata/processing/data_processor.py [hunk 1]
```diff
@@ -30,16 +30,16 @@
 from typing import Any, Dict, List, Optional, Tuple, TypeVar, Union
 from urllib import parse
 
+import boto3
+import botocore
```
Comment (awaelchli) [inline src/litdata/processing/data_processor.py:34]: Note that boto3 (including botocore) are core dependencies (requirements.txt) so we don't need a conditional import as before.

File: tests/streaming/test_resolver.py [hunk 2]
```diff
@@ -35,7 +35,7 @@ def test_src_resolver_s3_connections(monkeypatch):
 
     client_cls_mock = mock.MagicMock()
     client_cls_mock.return_value = client_mock
-    resolver.LightningClient = client_cls_mock
```
Comment (awaelchli) [inline tests/streaming/test_resolver.py:38]: I think technically this was a bug. Mocking things on the global modules without using monkeypatching could result in hard to debug test interactions. This might be done in more places of the test suite, so that could give problems in the future.

Comment (tchaton) [reply tests/streaming/test_resolver.py:38]: True. This should have been with monkeypatch

**Review by tchaton — APPROVED** [review]
